File: duplex/hs/src/Nightseam/Duplex/Wire.hs
{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE OverloadedStrings #-}
-- | Composition using the upstream Bitwire contracts directly.
module Nightseam.Duplex.Wire (WireError(..), at, mount, encodePath, decodePath) where

import Bitwire
import Control.Concurrent.STM
import Control.Exception (Exception, mask, onException, throwIO)
import Control.Monad (forM_, unless, when)
import qualified Data.ByteString as BS
import qualified Data.Map.Strict as Map
import qualified Data.Set as Set
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import Data.Typeable (Typeable)

data WireError = InvalidPath | NoRoute | ReceiverExists | WireClosed
  deriving (Eq, Show, Typeable)

instance Exception WireError

-- | Encode opaque Unicode scalar segments by their UTF-8 byte lengths.
encodePath :: Path -> Either WireError Text
encodePath path
  | any (T.any (\c -> c >= '\xD800' && c <= '\xDFFF')) path = Left InvalidPath
  | otherwise = Right (T.concat (map segment path))
  where segment value = T.pack (show (BS.length (TE.encodeUtf8 value))) <> ":" <> value

-- | Accept only the canonical encoding. No slash, dot or Unicode normalization
-- is performed. Decode each segment separately so a split code point is refused.
decodePath :: Text -> Either WireError Path
decodePath = go . TE.encodeUtf8
  where
    go input
      | BS.null input = Right []
      | otherwise = do
          let (digits, suffix) = BS.break (== 58) input
          when (BS.null digits || BS.null suffix || BS.any (\c -> c < 48 || c > 57) digits
                || (BS.length digits > 1 && BS.head digits == 48)) (Left InvalidPath)
          let rest = BS.tail suffix
              maximumLength = BS.length rest
              maximumDigits = BS.length (TE.encodeUtf8 (T.pack (show maximumLength)))
          -- Refuse oversized declarations before parsing, including arbitrarily
          -- large integers, and keep all arithmetic inside a bounded Int.
          when (BS.length digits > maximumDigits) (Left InvalidPath)
          let lengthInteger = BS.foldl' (\n c -> n * 10 + fromIntegral (c - 48)) (0 :: Integer) digits
          when (lengthInteger > fromIntegral maximumLength) (Left InvalidPath)
          let (bytes, remaining) = BS.splitAt (fromIntegral lengthInteger) rest
          value <- either (const (Left InvalidPath)) Right (TE.decodeUtf8' bytes)
          (value :) <$> go remaining

-- | A pure send-only path view; lifecycle remains with its endpoint owner.
at :: Wire -> Path -> Wire
at root prefix = Wire $ \path -> send root (prefix ++ path)

data Registration = Registration
  { registrationID :: Integer
  , registrationReceiver :: Receiver
  , registrationActive :: TVar Bool
  , registrationDetaches :: TVar [IO ()]
  }

data MountState = MountState
  { mountClosed :: Bool
  , mountNext :: Integer
  , mountRegistrations :: Map.Map Integer Registration
  }

-- | Mount consumes one segment. Empty keys are valid, while the origin itself
-- has no leaf. Closure detaches registrations and leaves borrowed children open.
mount :: Map.Map Text Endpoint -> IO Endpoint
mount children = do
  state <- newTVarIO (MountState False 0 Map.empty)
  let destination path = do
        either throwIO (const (pure ())) (encodePath path)
        atomically $ do
          current <- readTVar state
          when (mountClosed current) (throwSTM WireClosed)
          case path of
            key : rest -> maybe (throwSTM NoRoute) (\child -> pure (key, rest, child)) (Map.lookup key children)
            [] -> throwSTM NoRoute
      reserve receiver = atomically $ do
        current <- readTVar state
        when (mountClosed current) (throwSTM WireClosed)
        unless (Map.null (mountRegistrations current)) (throwSTM ReceiverExists)
        active <- newTVar True
        detaches <- newTVar []
        let identifier = mountNext current
            registration = Registration identifier receiver active detaches
        writeTVar state current
          { mountNext = identifier + 1
          , mountRegistrations = Map.insert identifier registration (mountRegistrations current)
          }
        pure registration
      remove registration notify code reason = do
        held <- atomically $ do
          active <- readTVar (registrationActive registration)
          if not active then pure Nothing else do
            writeTVar (registrationActive registration) False
            modifyTVar' state (\s -> s {mountRegistrations = Map.delete (registrationID registration) (mountRegistrations s)})
            detaches <- readTVar (registrationDetaches registration)
            writeTVar (registrationDetaches registration) []
            pure (Just detaches)
        case held of
          Nothing -> pure ()
          Just detaches -> do
            sequence_ detaches
            when notify (maybe (pure ()) (\f -> f code reason) (onClosed (registrationReceiver registration)))
      attach registration child receiver = mask $ \restore -> do
        detach <- restore (receive child receiver)
        active <- atomically $ do
          live <- readTVar (registrationActive registration)
          when live (modifyTVar' (registrationDetaches registration) (detach :))
          pure live
        unless active (detach >> throwIO WireClosed)
      receiveNamespace receiver = mask $ \restore -> do
        registration <- reserve receiver
        remaining <- newTVarIO (Map.keysSet children)
        let forget = remove registration False (Code 0) ""
            childEnded key code reason = do
              lastChild <- atomically $ do
                old <- readTVar remaining
                let next = Set.delete key old
                writeTVar remaining next
                pure (Set.member key old && Set.null next)
              when lastChild (remove registration True code reason)
            install (key, child) = attach registration child receiver
              { onMessage = (\f delivered -> f (key : delivered)) <$> onMessage receiver
              , onClosed = Just (childEnded key)
              }
        restore (mapM_ install (Map.toList children)) `onException` forget
        pure forget
      end code reason = mask $ \_ -> do
        registrations <- atomically $ do
          current <- readTVar state
          if mountClosed current then pure [] else do
            let held = Map.elems (mountRegistrations current)
            writeTVar state current {mountClosed = True}
            pure held
        forM_ registrations (\registration -> remove registration True code reason)
  pure Endpoint
    { endpointWire = Wire $ \path message -> do
        (_, rest, child) <- destination path
        send (endpointWire child) rest message
    , receive = receiveNamespace
    , close = end
    }

File: duplex/py/nightseam/duplex/wire.py
"""Relative-path bitwire.Wire contracts and pure selection and mount views."""

from __future__ import annotations

from collections.abc import Callable, Mapping
from typing import Literal

import bitwire

__all__ = [
    "WireError",
    "encode_path",
    "decode_path",
    "at",
    "mount",
]

class WireError(Exception):
    def __init__(self, code: Literal["closed", "no_route", "receiver_exists", "invalid_path"]):
        self.code = code
        super().__init__(f"bitwire.Wire {code}.")


def _utf8(value: str) -> bytes:
    try:
        return value.encode("utf-8")
    except UnicodeEncodeError:
        raise WireError("invalid_path") from None


def encode_path(path: bitwire.Path) -> str:
    """UTF-8 byte-length-prefixed scalar segments: [] is '', [''] is '0:'."""
    return "".join(f"{len(_utf8(segment))}:{segment}" for segment in path)


def decode_path(encoded: str) -> list[str]:
    """Accept canonical encoding only, preserving empty segments and BOMs."""
    data = _utf8(encoded)
    path = []
    offset = 0
    while offset < len(data):
        start, length = offset, 0
        while offset < len(data) and data[offset] != 58:
            digit = data[offset] - 48
            offset += 1
            if digit < 0 or digit > 9:
                raise WireError("invalid_path")
            length = length * 10 + digit
            if length > 9007199254740991:
                raise WireError("invalid_path")
        if offset == start or offset == len(data) or (offset - start > 1 and data[start] == 48):
            raise WireError("invalid_path")
        offset += 1
        if length > len(data) - offset:
            raise WireError("invalid_path")
        try:
            path.append(data[offset : offset + length].decode("utf-8"))
        except UnicodeDecodeError:
            raise WireError("invalid_path") from None
        offset += length
    return path


class _Selected:
    def __init__(self, root: bitwire.Wire, path: bitwire.Path):
        self._root, self._prefix = root, list(path)

    def send(self, path: bitwire.Path, message: bitwire.Message) -> None:
        self._root.send([*self._prefix, *path], message)


def at(root: bitwire.Wire, path: bitwire.Path) -> bitwire.Wire:
    """Select send access; receiving and closure remain with the endpoint owner."""
    encode_path(path)
    return _Selected(root, path)


class _Mounted:
    def __init__(self, children: Mapping[str, bitwire.Endpoint]):
        self._routes = dict(children)
        self._attachment = None
        self._closed = False

    def send(self, path: bitwire.Path, message: bitwire.Message) -> None:
        if self._closed:
            raise WireError("closed")
        encode_path(path)
        if not path or path[0] not in self._routes:
            raise WireError("no_route")
        self._routes[path[0]].send(path[1:], message)

    def receive(self, receiver: bitwire.Receiver) -> Callable[[], None]:
        if self._closed:
            raise WireError("closed")
        if self._attachment is not None:
            raise WireError("receiver_exists")
        detaches = []
        attachment = object()
        self._attachment = attachment, receiver, detaches
        remaining = set(self._routes)

        def detach():
            if self._attachment is None or self._attachment[0] is not attachment:
                return
            self._attachment = None
            for stop in detaches:
                stop()
            detaches.clear()

        def ended(key, code, reason):
            remaining.discard(key)
            if not remaining and self._attachment is not None and self._attachment[0] is attachment:
                detach()
                if receiver.closed:
                    receiver.closed(code, reason)

        try:
            for key, child in self._routes.items():
                def delivered(path, message, key=key):
                    if receiver.message:
                        return receiver.message([key, *path], message)
                stop = child.receive(bitwire.Receiver(message=delivered,
                    closed=lambda code, reason, key=key: ended(key, code, reason)))
                if self._attachment is None:
                    stop()
                    raise WireError("closed")
                detaches.append(stop)
        except BaseException:
            detach()
            raise
        return detach

    def close(self, code: int = 1000, reason: str = "") -> None:
        if self._closed:
            return
        self._closed = True
        held, self._attachment = self._attachment, None
        if held is not None:
            _, receiver, detaches = held
            for detach in detaches:
                detach()
            if receiver.closed:
                receiver.closed(code, reason)


def mount(children: Mapping[str, bitwire.Endpoint]) -> bitwire.Endpoint:
    """Mount borrowed endpoints; closing releases only this mount's attachments."""
    return _Mounted(children)
